plugins/Diagnostics/ConfigReader.php

```diff
@@ -0,0 +1,193 @@
+use Piwik\Settings as PiwikSettings;
+use Piwik\Plugin\Settings as PluginSettings;
+
+/**
+ * A diagnostic report contains all the results of all the diagnostics.
+ */
+class ConfigReader
+{
+    /**
+     * @var GlobalSettingsProvider
+     */
+    private $settings;
+
+    /**
+     * @var IniReader
+     */
+    private $iniReader;
+
+    public function __construct(GlobalSettingsProvider $settings, IniReader $iniReader)
+    {
+        $this->settings = $settings;
+        $this->iniReader = $iniReader;
+    }
+
+    public function getConfigValuesFromFiles()
+    {
+        $ini = $this->settings->getIniFileChain();
+        $descriptions = $this->iniReader->readComments($this->settings->getPathGlobal());
+
+        $copy = array();
+        foreach ($ini->getAll() as $category => $values) {
+            if ($this->shouldSkipCategory($category)) {
+                continue;
+            }
+
+            $local = $this->getFromLocalConfig($category);
+            if (empty($local)) {
+                $local = array();
+            }
+
+            $global = $this->getFromGlobalConfig($category);
+            if (empty($global)) {
+                $global = array();
+            }
+
+            $copy[$category] = array();
+            foreach ($values as $key => $value) {
+
+                $newValue = $value;
+                if ($this->isKeyAPassword($key)) {
+                    $newValue = $this->maskPassword($value);
+                }
+
+                $defaultValue = null;
+                if (array_key_exists($key, $global)) {
+                    $defaultValue = $global[$key];
+                }
+
+                $description = '';
+                if (!empty($descriptions[$category][$key])) {
+                    $description = trim($descriptions[$category][$key]);
+                }
+
+                $copy[$category][$key] = array(
+                    'value' => $newValue,
+                    'description' => $description,
+                    'isCustomValue' => array_key_exists($key, $local),
+                    'defaultValue' => $defaultValue,
+                );
+            }
+        }
+
+        return $copy;
+    }
+
+    private function shouldSkipCategory($category)
+    {
+        $category = strtolower($category);
+        if ($category === 'database') {
+            return true;
+        }
+
+        $developmentOnlySections = array('database_tests', 'tests', 'debugtests');
+
+        return !Development::isEnabled() && in_array($category, $developmentOnlySections);
+    }
+
+    public function getFromGlobalConfig($name)
+    {
+        return $this->settings->getIniFileChain()->getFrom($this->settings->getPathGlobal(), $name);
+    }
+
+    public function getFromLocalConfig($name)
+    {
+        return $this->settings->getIniFileChain()->getFrom($this->settings->getPathLocal(), $name);
+    }
+
+    private function maskPassword($password)
+    {
+        return str_pad('', strlen($password), '*');
```

Comment:
Not sure if we should "show" the length of a password by replacing each char with a `*`. Maybe it would be easier and safer by simply always showing the same amount of stars. Any other opinions on that?